File: mutation-tests-ts-example/DiscountCalculator.test.ts
import { DiscountCalculator } from './DiscountCalculator';

describe('DiscountCalculator', () => {
    let calculator: DiscountCalculator;

    beforeEach(() => {
        calculator = new DiscountCalculator();
    });

    test('should throw error for non-positive price', () => {
        expect(() => calculator.calculateFinalPrice(0, 'bronze', 1, false))
            .toThrow('Base price must be positive');
    });

    test('should apply bronze tier discount', () => {
        const result = calculator.calculateFinalPrice(100, 'bronze', 1, false);
        expect(result).toBe(95); // 5% off
    });

    test('should apply gold tier discount', () => {
        const result = calculator.calculateFinalPrice(100, 'gold', 1, false);
        expect(result).toBe(85); // 15% off
    });

    test('should apply bulk discount', () => {
        const result = calculator.calculateFinalPrice(100, 'bronze', 10, false);
        expect(result).toBe(90); // 5% tier + 5% bulk = 10% total
    });

    test('should handle unknown tier', () => {
        const result = calculator.calculateFinalPrice(100, 'unknown', 1, false);
        expect(result).toBe(100); // no discount
    });
});

File: mutation-tests-java-example/src/test/java/com/github/mslowiak/pricing/DiscountCalculatorTest.java
package com.github.mslowiak.pricing;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class DiscountCalculatorTest {

    private DiscountCalculator calculator;

    @BeforeEach
    void setUp() {
        calculator = new DiscountCalculator();
    }

    @Test
    void shouldThrowExceptionForNonPositivePrice() {
        assertThrows(IllegalArgumentException.class,
                () -> calculator.calculateFinalPrice(0, "bronze", 1, false));
    }

    @Test
    void shouldApplyBronzeTierDiscount() {
        double result = calculator.calculateFinalPrice(100, "bronze", 1, false);
        assertEquals(95.0, result, 0.01); // 5% off
    }

    @Test
    void shouldApplyGoldTierDiscount() {
        double result = calculator.calculateFinalPrice(100, "gold", 1, false);
        assertEquals(85.0, result, 0.01); // 15% off
    }

    @Test
    void shouldApplyBulkDiscount() {
        double result = calculator.calculateFinalPrice(100, "bronze", 10, false);
        assertEquals(90.0, result, 0.01); // 5% tier + 5% bulk = 10% total
    }

    @Test
    void shouldHandleUnknownTier() {
        double result = calculator.calculateFinalPrice(100, "unknown", 1, false);
        assertEquals(100.0, result, 0.01); // no discount
    }
}